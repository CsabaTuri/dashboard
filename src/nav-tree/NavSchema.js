import React from 'react'
import { Nav } from 'office-ui-fabric-react'
import { browserHistory } from 'react-router';

import { connect } from 'react-redux'

function navLinkClicked(e, link) {
  e.preventDefault();
  browserHistory.push(link.url)
}

function NavSchema({splat, classes, indexes}) {
  const dbpath = splat ? `${splat}/` : ""

  const options = {
    name : "Options",
    links : [
      {
        name : "Create Database",
        key : `${dbpath}databases`,
        url : `/db/${dbpath}databases`
      },
      {
        name : "Create Class",
        key : `${dbpath}classes`,
        url : `/db/${dbpath}classes`
      },
      {
        name : "Create Index",
        key : `${dbpath}indexes`,
        url : `/db/${dbpath}indexes`
      }
    ]
  }

  return <Nav groups={[options, classes, indexes]} onLinkClick={navLinkClicked} />
}

function mapStateToProp(state, ownProps) {
  const dbpath = ownProps.splat;

  const indexesLinks = !state.indexes.byName ? [] :
    Object.keys(state.indexes.byName)
      .map(key => {
        const slug = dbpath ? dbpath+"/indexes" : "indexes"
        const index = state.indexes.byName[key].indexInfo
        const name = index.ref.id

        return {
          name : name,
          url : "/db/"+slug+"/"+name,
          key : slug+"/"+name
        }
      })

  const classesLinks = !state.classes.byName ? [] :
    Object.keys(state.classes.byName)
      .map(key => {
        const slug = dbpath ? dbpath+"/classes" : "classes"
        const clazz = state.classes.byName[key].classInfo
        const name = clazz.ref.id

        return {
          name : name,
          url : "/db/"+slug+"/"+name,
          key : slug+"/"+name
        }
      })

  return {
    classes: {
      name: "classes",
      links: classesLinks
    },
    indexes: {
      name: "indexes",
      links: indexesLinks
    }
  }
}

export default connect(
  mapStateToProp
)(NavSchema)
