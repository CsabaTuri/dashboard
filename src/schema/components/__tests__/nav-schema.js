import React from "react"
import { Map, List } from "immutable"
import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"

import NavSchema from "../nav-schema"
import { reduceSchemaTree } from "../../"

jest.mock("react-router", () => ({
  browserHistory: {
    push: jest.fn()
  }
}))

const { browserHistory } = require("react-router")

describe("NavSchema Component", () => {
  let rootDb, subDb, store, client

  const render = (db) => shallow(
    <NavSchema
      store={store}
      client={client}
      database={db} />
  ).dive()

  beforeEach(() => {
    rootDb = Map.of(
      "isRoot", true,
      "url", "/db/databases",
      "classes", List.of(Map.of("name", "a-class", "url", "/db/classes/a-class")),
      "indexes", List.of(Map.of("name", "a-index", "url", "/db/indexes/a-index")),
    )

    subDb = Map.of(
      "isRoot", false,
      "url", "/db/subdb/databases",
      "classes", List(),
      "indexes", List(),
    )

    client = { hasPrivileges: jest.fn(() => true) }
    store = createImmutableTestStore({ schema: reduceSchemaTree })()
  })

  it("should render classes and indexes for root db", () => {
    expect(shallowToJson(render(rootDb))).toMatchSnapshot()
  })

  it("should render classes and indexes for sub db", () => {
    expect(shallowToJson(render(subDb))).toMatchSnapshot()
  })

  it("should render options for root using a server key", () => {
    client.hasPrivileges.mockReturnValue(false)
    expect(shallowToJson(render(rootDb))).toMatchSnapshot()
  })

  it("should render options for sub using a server key", () => {
    client.hasPrivileges.mockReturnValue(false)
    expect(shallowToJson(render(subDb))).toMatchSnapshot()
  })

  it("navigates to the clicked resource", () => {
    const comp = render(rootDb)
    comp.simulate("linkClick", new Event("click"), { url: "resource-url" })
    expect(browserHistory.push).toHaveBeenCalledWith("resource-url")
  })
})
