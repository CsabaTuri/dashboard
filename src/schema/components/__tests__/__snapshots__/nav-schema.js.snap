// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavSchema Component should hide admin options when using a server key 1`] = `
<CustomNav
  groups={
    Array [
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "create-class",
            "name": "Create Class",
            "url": "/db/classes",
          },
          Object {
            "key": "create-index",
            "name": "Create Index",
            "url": "/db/indexes",
          },
        ],
        "name": "Options",
      },
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "/db/classes/a-class",
            "name": "a-class",
            "url": "/db/classes/a-class",
          },
        ],
        "name": "Classes",
      },
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "/db/indexes/a-index",
            "name": "a-index",
            "url": "/db/indexes/a-index",
          },
        ],
        "name": "Indexes",
      },
    ]
  }
  onLinkClick={[Function]}
  onRenderLink={[Function]}
/>
`;

exports[`NavSchema Component should render classes and indexes 1`] = `
<CustomNav
  groups={
    Array [
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "create-database",
            "name": "Create Database",
            "url": "/db/databases",
          },
          Object {
            "key": "create-class",
            "name": "Create Class",
            "url": "/db/classes",
          },
          Object {
            "key": "create-index",
            "name": "Create Index",
            "url": "/db/indexes",
          },
          Object {
            "key": "manage-keys",
            "name": "Manage Keys",
            "url": "/db/keys",
          },
        ],
        "name": "Options",
      },
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "/db/classes/a-class",
            "name": "a-class",
            "url": "/db/classes/a-class",
          },
        ],
        "name": "Classes",
      },
      Object {
        "isExpanded": true,
        "links": Array [
          Object {
            "key": "/db/indexes/a-index",
            "name": "a-index",
            "url": "/db/indexes/a-index",
          },
        ],
        "name": "Indexes",
      },
    ]
  }
  onLinkClick={[Function]}
  onRenderLink={[Function]}
/>
`;
